### DISCUSSION TITLE
Question about model-based testing in x-state with mocked services

### DISCUSSION THREAD

**Discussion by bhardy** [statelyai/xstate#1962] [Q&A]
We're trying to implement Model-Based testing on our purchase flows. We're running into issues with mocked services. I've made a base case in codesandbox. Please ignore the actual app, we're just looking at the test right now: https://codesandbox.io/embed/x-state-model-based-testing-with-services-7ugly?file=/src/buyMachine.test.js (check the file `buyMachine.test.js`)

The problem we're facing is that the first invocation of the service always seems to pass but subsequent ones seem to result in the machine being stuck on the previous successful (invoke) state. This is the absolute base case I could find to recreate the issue. You'll notice if you swap lines 27 & 28, the error case will pass and the success case will fail. In our app, we have this issue but we also have an issue with testing an additional service in the next step (if we force a happy path). I assume both are the same issue.

I assume we're just doing something wrong (maybe with our promise resolve/reject) and this isn't an x-state problem but I'll be happy to make a PR to the docs if solving this pattern helps people in the future.

**Comment by davidkpiano** [comment]
Looking at your test code, there is nothing in the machine itself invoking the services (which is good), so I can almost guarantee this is a problem outside of XState. Probably some incorrect caching?

**Reply by bhardy** [reply]
@davidkpiano yea… my understanding is that the test-machine should not invoke the services but it should have actions that handle the possible results from the invoke steps of the real-machine. It seems to me like something is hung up in the promise resolution but a few of us can't figure out if we've made a mistake with our testing configs or if our mocked promises are being accidentally reused or something like that. Do you know of any examples of someone doing model-based testing with invoked services?

**Reply by bhardy** [reply]
OK, we got this working with promises. It was a silly little mistake. https://codesandbox.io/s/x-state-model-based-testing-with-services-forked-px7hp?file=/src/buyMachine.test.js

Next, we'll try to pull the promises out because they do seem unnecessary.

**Reply by bhardy** [reply]
Here it is with jest mocking but no promises. Let me know if you think this example could help in the docs. Special thanks to @jimwheaton for helping with this. https://codesandbox.io/s/x-state-model-based-testing-with-services-no-promises-nm1pv

**Comment by bhardy** [comment]
It turned out that our original sandbox had a little mistake which we fixed here: https://codesandbox.io/s/x-state-model-based-testing-with-services-forked-px7hp?file=/src/buyMachine.test.js

We ended up removing the promises, which may be a better example here: https://codesandbox.io/s/x-state-model-based-testing-with-services-no-promises-nm1pv?file=/src/buyMachine.test.js

**Comment by seloner** [comment]
Nice example I suggest adding it to the docs
